rmw_connext_shared_cpp/src/shared_functions.cpp

```diff
@@ -186,6 +241,53 @@ get_datawriter_qos(DDSDomainParticipant * participant, DDS_DataWriterQos & dataw
     RMW_SET_ERROR_MSG("failed to add qos property");
     return false;
   }
+  switch (qos_profile.history) {
+    case RMW_QOS_POLICY_KEEP_LAST_HISTORY:
+      datawriter_qos.history.kind = DDS_KEEP_LAST_HISTORY_QOS;
+      break;
+    case RMW_QOS_POLICY_KEEP_ALL_HISTORY:
+      datawriter_qos.history.kind = DDS_KEEP_ALL_HISTORY_QOS;
+      break;
+    case RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT:
+      break;
+    default:
+      RMW_SET_ERROR_MSG("Unknown QoS history policy");
+      return false;
+  }
+
+  switch (qos_profile.reliability) {
+    case RMW_QOS_POLICY_BEST_EFFORT:
+      datawriter_qos.reliability.kind = DDS_BEST_EFFORT_RELIABILITY_QOS;
+      break;
+    case RMW_QOS_POLICY_RELIABLE:
+      datawriter_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
+      break;
+    case RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT:
+      break;
+    default:
+      RMW_SET_ERROR_MSG("Unknown QoS reliability policy");
+      return false;
+  }
+
+  if (qos_profile.depth != RMW_QOS_POLICY_DEPTH_SYSTEM_DEFAULT) {
+    datawriter_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+  }
+
+  // ensure the history depth is at least the requested queue size
+  assert(datawriter_qos.history.depth >= 0);
+  if (
+    datawriter_qos.history.kind == DDS::KEEP_LAST_HISTORY_QOS &&
+    static_cast<size_t>(datawriter_qos.history.depth) < qos_profile.depth
+  )
+  {
+    if (qos_profile.depth > (std::numeric_limits<DDS_Long>::max)()) {
+      RMW_SET_ERROR_MSG(
+        "failed to set history depth since the requested queue size exceeds the DDS type");
+      return false;
+    }
+    datawriter_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+  }
+
```

Comment:
Done. Sharing this code also forces the DataReader and DataWriter to interpret QoS profiles in the same way, which is probably what we want.